_posts/2024-01-22-Python-Memory-Management-and-Troubleshooting.md

```diff
@@ -0,0 +1,568 @@
+```
+
+</div>
+
+_Memory heaps help organize available memory on a computer for specific procedures. Heaps may have one or many memory pools._
+
+Computer memory blocks may be organized in hierarchical layers to manage memory efficiently or towards a specific purpose.
+One top-level organization model for computer memory is through the use of ___heaps___ which help describe chunks of the total memory available on a computer for specific processes.
+These heaps may be ___private___ (only available to a specific software process) or ___shared___ (available to one or many software processes).
+Heaps are sometimes further segmented into ___pools___ which are areas of the heap which can be used for specific purposes.
+
+### Memory Allocator
+
+```mermaid!
+sequenceDiagram
+
+software ->> allocator:"I need some memory, please!"
+allocator ->> memory:Allocate portion of  memory for use
+memory ->> allocator:Allocated memory for use
+allocator ->> software:"Here's some memory to use!"
+
+software ->> allocator:"I'm finished with that memory!"
+allocator ->> memory:Free the memory for other purposes.
+```
+
+_Memory allocators help software reserve and free computer memory resources._
+
+Memory management is a concept which helps enable the shared use of computer memory to avoid challenges such as memory overuse (where all memory is in use and never shared to other software).
+Computer memory management often occurs through the use of a ___memory allocator___ which controls how computer memory resources are used for software.
+Computer software is written to interact with memory allocators to use computer memory.
+Memory allocators may be used manually (with specific directions provided on when and how to use memory resources) or automatically (with an algorithmic approach of some kind).
+The memory allocator usually performs the following actions with memory (in addition to others):
+
+- __"Allocation"__: computer memory resource reservation (taking memory). This is sometimes also known as "`alloc`", or "allocate memory".
+- __"Deallocation"__: computer memory resource freeing (giving back memory for other uses). This is sometimes also known as "`free`", or "freeing memory from allocation".
+
+### Garbage Collection
+
+```mermaid!
+sequenceDiagram
+
+participant software
+participant memory allocator
+participant garbage collector
+participant memory
+
+software ->> memory allocator:"I need some memory, please!"
+memory allocator ->> memory:Allocate portion of  memory for use
+memory ->> memory allocator:Allocated memory for use
+memory allocator ->> software:"Here's some memory to use!"
+
+software -->> garbage collector :(software finishes with memory usage, sometimes implicitly)
+garbage collector  ->> memory:Free the memory for other purposes.
+```
+
+_Garbage collectors help free computer memory which is no longer referenced by software._
+
+"Garbage collection (GC)" is used to describe a type of automated memory management.
```

Comment:
Maybe throw in a quick sentence about the benefits/motivation of GC, e.g. reduce human error, avoid common critical point of failure. Unless you talk more about this below. Still reading.

Reply:
Thanks for this, great idea! I'll add this in.